components/script/dom/console.rs

```diff
@@ -55,14 +58,39 @@ where
     f()
 }
 
-fn console_messages(global: &GlobalScope, messages: &[DOMString], level: LogLevel) {
-    console_message(global, DOMString::from(messages.join(" ")), level)
+#[allow(unsafe_code)]
+fn stringify_message(message: HandleValue) -> DOMString {
+    let cx = *GlobalScope::get_cx();
+    unsafe {
+        let js_string = if message.is_string() {
+            // directly print string
+            message.to_string()
+        } else if message.is_undefined() {
+            // better value than "(void 0)" from JS_ValueToSource
+            return DOMString::from("undefined");
+        } else {
+            jsapi::JS_ValueToSource(cx, message.into())
+        };
```

Comment:
The values produced are a bit odd (for instance the way booleans) are represented. Could this do `JSON.stringify()` instead?

Reply:
I moved some of the logic for producing the values into Servo, which improved the output in a few cases (e.g. for `undefined` in an object).

Note that Booleans are represented as `true`/`false` with `JS_ValueToSource` - the `new Boolean(...)` display is only used for the [Boolean primitive wrapper objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean/Boolean#return_value), which are essentially never used.